### DISCUSSION TITLE
Calculating inbreeding coefficients from pedigrees

### DISCUSSION THREAD

**Discussion by JimmyPeniston** [tskit-dev/tskit#2711] [Q&A] [answered]
Hi all,

I have tree sequence data from a pretty simple non-WF SLIM simulation and I'm wondering if anyone has advice for a straightforward way to calculate pedigree-based inbreeding coefficients. The best method I can think of is to try building some sort of recursive algorithm that loops through individuals.parents, but I'm not sure if there is an easier/more efficient way.

For context, I am currently only interested in pedigree-based metrics. So, my chromosome length is set to 1 and a mutation rate of 0.

Thank you for your help. Sorry if this is a silly question or the wrong place to post this. I'm very new to using tskit.

**Comment by jeromekelleher** [comment, marked as answer]
Hi @JimmyPeniston :wave:

We don't have any code for directly computing pedigree statistics in tskit, but there are some very powerful methods in [sgkit](https://pystatgen.github.io/sgkit/latest/index.html) like [pedigree_inbreeding](https://pystatgen.github.io/sgkit/latest/generated/sgkit.pedigree_inbreeding.html) which will hopefully do what you need.

The sgkit code has just been released and I don't think anyone has linked up simulations with it yet. I guess the main thing is whether SLiM saves the pedigree information in the individual table (which I'm not sure about). 

Can you run a very small example, and print out the contents of the individual and node tables here please (e.g. ``print(ts.tables.nodes)``)? Hopefully hooking the tskit data to sgkit will then be fairly straightforward.

This is probably quite a useful thing in general, so we should document how to do it somewhere, I guess (assuming it works).

**Reply by JimmyPeniston** [reply]
Thanks for the response! Here are the individual and node tables from a small example.

Individual Table
╔═╤════╤═══════╤═══╤════════════════════════════════════════╗
║id │ flags  │    location  │parents│       metadata                                        ║
╠═╪════╪═══════╪═══╪═════════════════════════╣
║0 │65536│0.0, 0.0, 0.0│ -1, -1│{'pedigree_id': 35398, 'pedigree_p1':...║
║1 │65536│0.0, 0.0, 0.0│ -1, -1│{'pedigree_id': 35400, 'pedigree_p1':...║
║2 │65536│0.0, 0.0, 0.0│ -1, -1│{'pedigree_id': 35411, 'pedigree_p1':...║
║3 │65536│0.0, 0.0, 0.0│ -1, -1│{'pedigree_id': 35412, 'pedigree_p1':...║
║4 │65536│0.0, 0.0, 0.0│ -1, -1│{'pedigree_id': 35433, 'pedigree_p1':...║
║5 │65536│0.0, 0.0, 0.0│ -1, -1│{'pedigree_id': 35458, 'pedigree_p1':...║
║6 │65536│0.0, 0.0, 0.0│ -1, 17│{'pedigree_id': 35485, 'pedigree_p1':...║
║7 │65536│0.0, 0.0, 0.0│  -1, 1│{'pedigree_id': 35489, 'pedigree_p1':...║
║8 │65536│0.0, 0.0, 0.0│ -1, -1│{'pedigree_id': 35490, 'pedigree_p1':...║
║9 │65536│0.0, 0.0, 0.0│  0, -1│{'pedigree_id': 35500, 'pedigree_p1':...║
║10│65536│0.0, 0.0, 0.0│ -1, -1│{'pedigree_id': 35504, 'pedigree_p1':...║
╚═╧════╧═══════╧═══╧════════════════════════════════════════╝

Node Table
╔═╤═════╤══════════╤══════════╤════╤════════════════════════════════════════╗
║id │   flags   │population          │individual            │   time  │          metadata                                     
╠═╪═════╪══════════╪══════════╪════╪════════════════════════════════════════╣
║0 │             1│                          1│                          0│          3│     {'slim_id': 70796, 'is_null': False, ...║
║1 │             1│                           1│                         0│           3│     {'slim_id': 70797, 'is_null': False, ...║
║2 │            1│                           1│                          1│           3│     {'slim_id': 70800, 'is_null': False, ...║
║3 │            1│                           1│                          1│           3│     {'slim_id': 70801, 'is_null': False, ...║
║4 │            1│                          2│                          2│           3│     {'slim_id': 70822, 'is_null': False, ...║
║5 │            1│                          2│                          2│           3│     {'slim_id': 70823, 'is_null': False, ...║
║6 │            1│                          2│                          3│           3│     {'slim_id': 70824, 'is_null': False, ...║
║7 │            1│                          2│                          3│           3│     {'slim_id': 70825, 'is_null': False, ...║
║8 │            1│                          1│                          4│           2│     {'slim_id': 70866, 'is_null': False, ...║
║9 │            1│                          1│                          4│           2│     {'slim_id': 70867, 'is_null': False, ...║
║10│           1│                          1│                          5│            1│     {'slim_id': 70916, 'is_null': False, ...║
║11│           1│                           1│                          5│            1│     {'slim_id': 70917, 'is_null': False, ...║
║12│          1│                           2│                          6│            1│     {'slim_id': 70970, 'is_null': False, ...║
║13│          1│                           2│                          6│            1│     {'slim_id': 70971, 'is_null': False, ...║
║14│          1│                           1│                          7│            0│     {'slim_id': 70978, 'is_null': False, ...║
║15│          1│                           1│                          7│            0│     {'slim_id': 70979, 'is_null': False, ...║
║16│          1│                           1│                          8│            0│     {'slim_id': 70980, 'is_null': False, ...║
║17│          1│                           1│                          8│            0│     {'slim_id': 70981, 'is_null': False, ...║
║18│          1│                           1│                          9│            0│     {'slim_id': 71000, 'is_null': False, ...║
║19│          1│                           1│                          9│            0│     {'slim_id': 71001, 'is_null': False, ...║
╚══╧═════╧══════════╧══════════╧════╧════════════════════════════════════════╝

**Reply by jeromekelleher** [reply]
Hmm, looks like it's only keeping some of the pedigree individuals (notice individual 6 has a parent 17, but this individual doesn't exist).

Is there an option to get SLiM to fully populate the individual table?

**Reply by JimmyPeniston** [reply]
Opps, sorry. I cut off the tables when copy and pasting them. Here's the full output.  I'm still not sure if it is a problem that a bunch of the parents are "-1". I think it might be only keeping track of parents that are still alive at the end of the simulation. There might be some way to work around that if needed.

Individual Table
╔══╤═════╤═════════════╤═══════╤════════════════════════════════════════╗
║id│flags│location     │parents│metadata                                ║
╠══╪═════╪═════════════╪═══════╪════════════════════════════════════════╣
║0 │65536│0.0, 0.0, 0.0│ -1, -1│{'pedigree_id': 35398, 'pedigree_p1':...║
║1 │65536│0.0, 0.0, 0.0│ -1, -1│{'pedigree_id': 35400, 'pedigree_p1':...║
║2 │65536│0.0, 0.0, 0.0│ -1, -1│{'pedigree_id': 35411, 'pedigree_p1':...║
║3 │65536│0.0, 0.0, 0.0│ -1, -1│{'pedigree_id': 35412, 'pedigree_p1':...║
║4 │65536│0.0, 0.0, 0.0│ -1, -1│{'pedigree_id': 35433, 'pedigree_p1':...║
║5 │65536│0.0, 0.0, 0.0│ -1, -1│{'pedigree_id': 35458, 'pedigree_p1':...║
║6 │65536│0.0, 0.0, 0.0│ -1, 17│{'pedigree_id': 35485, 'pedigree_p1':...║
║7 │65536│0.0, 0.0, 0.0│  -1, 1│{'pedigree_id': 35489, 'pedigree_p1':...║
║8 │65536│0.0, 0.0, 0.0│ -1, -1│{'pedigree_id': 35490, 'pedigree_p1':...║
║9 │65536│0.0, 0.0, 0.0│  0, -1│{'pedigree_id': 35500, 'pedigree_p1':...║
║10│65536│0.0, 0.0, 0.0│ -1, -1│{'pedigree_id': 35504, 'pedigree_p1':...║
║11│65536│0.0, 0.0, 0.0│  -1, 2│{'pedigree_id': 35503, 'pedigree_p1':...║
║12│65536│0.0, 0.0, 0.0│ 25, -1│{'pedigree_id': 35511, 'pedigree_p1':...║
║13│65536│0.0, 0.0, 0.0│ -1, 16│{'pedigree_id': 35517, 'pedigree_p1':...║
║14│65536│0.0, 0.0, 0.0│ 26, 25│{'pedigree_id': 35519, 'pedigree_p1':...║
║15│65536│0.0, 0.0, 0.0│ -1, -1│{'pedigree_id': 35378, 'pedigree_p1':...║
║16│65536│0.0, 0.0, 0.0│ -1, -1│{'pedigree_id': 35377, 'pedigree_p1':...║
║17│65536│0.0, 0.0, 0.0│ -1, -1│{'pedigree_id': 35415, 'pedigree_p1':...║
║18│65536│0.0, 0.0, 0.0│ 16, -1│{'pedigree_id': 35439, 'pedigree_p1':...║
║19│65536│0.0, 0.0, 0.0│ -1, -1│{'pedigree_id': 35448, 'pedigree_p1':...║
║20│65536│0.0, 0.0, 0.0│ -1, -1│{'pedigree_id': 35444, 'pedigree_p1':...║
║21│65536│0.0, 0.0, 0.0│ -1, -1│{'pedigree_id': 35447, 'pedigree_p1':...║
║22│65536│0.0, 0.0, 0.0│ -1, -1│{'pedigree_id': 35471, 'pedigree_p1':...║
║23│65536│0.0, 0.0, 0.0│  -1, 3│{'pedigree_id': 35469, 'pedigree_p1':...║
║24│65536│0.0, 0.0, 0.0│ -1, -1│{'pedigree_id': 35463, 'pedigree_p1':...║
║25│65536│0.0, 0.0, 0.0│ -1, -1│{'pedigree_id': 35481, 'pedigree_p1':...║
║26│65536│0.0, 0.0, 0.0│ 21, 21│{'pedigree_id': 35487, 'pedigree_p1':...║
║27│65536│0.0, 0.0, 0.0│ 15, -1│{'pedigree_id': 35520, 'pedigree_p1':...║
║28│65536│0.0, 0.0, 0.0│ -1, -1│{'pedigree_id': 35514, 'pedigree_p1':...║
║29│65536│0.0, 0.0, 0.0│ -1, 16│{'pedigree_id': 35510, 'pedigree_p1':...║
║30│65536│0.0, 0.0, 0.0│ 17, -1│{'pedigree_id': 35515, 'pedigree_p1':...║
║31│65536│0.0, 0.0, 0.0│ 24, -1│{'pedigree_id': 35521, 'pedigree_p1':...║
╚══╧═════╧═════════════╧═══════╧════════════════════════════════════════╝

Node Table
╔══╤═════╤══════════╤══════════╤════╤════════════════════════════════════════╗
║id│flags│population│individual│time│metadata                                ║
╠══╪═════╪══════════╪══════════╪════╪════════════════════════════════════════╣
║0 │    1│         1│         0│   3│{'slim_id': 70796, 'is_null': False, ...║
║1 │    1│         1│         0│   3│{'slim_id': 70797, 'is_null': False, ...║
║2 │    1│         1│         1│   3│{'slim_id': 70800, 'is_null': False, ...║
║3 │    1│         1│         1│   3│{'slim_id': 70801, 'is_null': False, ...║
║4 │    1│         2│         2│   3│{'slim_id': 70822, 'is_null': False, ...║
║5 │    1│         2│         2│   3│{'slim_id': 70823, 'is_null': False, ...║
║6 │    1│         2│         3│   3│{'slim_id': 70824, 'is_null': False, ...║
║7 │    1│         2│         3│   3│{'slim_id': 70825, 'is_null': False, ...║
║8 │    1│         1│         4│   2│{'slim_id': 70866, 'is_null': False, ...║
║9 │    1│         1│         4│   2│{'slim_id': 70867, 'is_null': False, ...║
║10│    1│         1│         5│   1│{'slim_id': 70916, 'is_null': False, ...║
║11│    1│         1│         5│   1│{'slim_id': 70917, 'is_null': False, ...║
║12│    1│         2│         6│   1│{'slim_id': 70970, 'is_null': False, ...║
║13│    1│         2│         6│   1│{'slim_id': 70971, 'is_null': False, ...║
║14│    1│         1│         7│   0│{'slim_id': 70978, 'is_null': False, ...║
║15│    1│         1│         7│   0│{'slim_id': 70979, 'is_null': False, ...║
║16│    1│         1│         8│   0│{'slim_id': 70980, 'is_null': False, ...║
║17│    1│         1│         8│   0│{'slim_id': 70981, 'is_null': False, ...║
║18│    1│         1│         9│   0│{'slim_id': 71000, 'is_null': False, ...║
║19│    1│         1│         9│   0│{'slim_id': 71001, 'is_null': False, ...║
║                 58 rows skipped (tskit.set_print_options)                  ║
║78│    0│         1│        -1│   9│{'slim_id': 70307, 'is_null': False, ...║
║79│    0│         1│        -1│  10│{'slim_id': 70231, 'is_null': False, ...║
║80│    0│         2│        -1│  12│{'slim_id': 70116, 'is_null': False, ...║
║81│    0│         1│        -1│  13│{'slim_id': 70029, 'is_null': False, ...║
║82│    0│         1│        -1│  14│{'slim_id': 69951, 'is_null': False, ...║
║83│    0│         2│        -1│  15│{'slim_id': 69917, 'is_null': False, ...║
║84│    0│         1│        -1│  17│{'slim_id': 69750, 'is_null': False, ...║
║85│    0│         2│        -1│  17│{'slim_id': 69796, 'is_null': False, ...║
║86│    0│         1│        -1│  18│{'slim_id': 69700, 'is_null': False, ...║
║87│    0│         2│        -1│  23│{'slim_id': 69392, 'is_null': False, ...║
║88│    0│         2│        -1│  26│{'slim_id': 69192, 'is_null': False, ...║
║89│    0│         2│        -1│  32│{'slim_id': 68705, 'is_null': False, ...║
║90│    0│         1│        -1│  39│{'slim_id': 68194, 'is_null': False, ...║
║91│    0│         2│        -1│  42│{'slim_id': 67992, 'is_null': False, ...║
║92│    0│         2│        -1│  51│{'slim_id': 67380, 'is_null': False, ...║
║93│    0│         2│        -1│  51│{'slim_id': 67388, 'is_null': False, ...║
║94│    0│         2│        -1│  51│{'slim_id': 67401, 'is_null': False, ...║
║95│    0│         2│        -1│ 104│{'slim_id': 63505, 'is_null': False, ...║
║96│    0│         1│        -1│ 135│{'slim_id': 61416, 'is_null': False, ...║
║97│    0│         2│        -1│ 999│{'slim_id': 41, 'is_null': False, 'ge...║
╚══╧═════╧══════════╧══════════╧════╧════════════════════════════════════════╝

**Reply by petrelharp** [reply]
SLiM will record pedigree information for all individuals in the tree sequence at the end; by default this will not necessarily be everyone you need - see https://tskit.dev/pyslim/docs/stable/tutorial.html#historical-individuals for details. To get *everyone* you would need to do something like `treeSeqRememberIndividuals(sim.subpopulations.individuals)` every tick. (And, that works fine, but it's easy to get very big tree sequences here if you have a big simulation.)

**Reply by JimmyPeniston** [reply]
Thanks @petrelharp! @jeromekelleher below are the individual and node tables that now have all parents.

╔════╤══════╤═════════════╤══════════╤════════════════════════════════════════╗
║id  │flags │location     │parents   │metadata                                ║
╠════╪══════╪═════════════╪══════════╪════════════════════════════════════════╣
║0   │196608│0.0, 0.0, 0.0│9587, 9605│{'pedigree_id': 36067, 'pedigree_p1':...║
║1   │196608│0.0, 0.0, 0.0│9576, 9595│{'pedigree_id': 36084, 'pedigree_p1':...║
║2   │196608│0.0, 0.0, 0.0│9604, 9607│{'pedigree_id': 36073, 'pedigree_p1':...║
║3   │196608│0.0, 0.0, 0.0│9618, 9618│{'pedigree_id': 36102, 'pedigree_p1':...║
║4   │196608│0.0, 0.0, 0.0│9586, 9586│{'pedigree_id': 36104, 'pedigree_p1':...║
║5   │196608│0.0, 0.0, 0.0│9614, 9612│{'pedigree_id': 36134, 'pedigree_p1':...║
║6   │196608│0.0, 0.0, 0.0│9605, 9604│{'pedigree_id': 36114, 'pedigree_p1':...║
║7   │196608│0.0, 0.0, 0.0│9608, 9609│{'pedigree_id': 36150, 'pedigree_p1':...║
║8   │196608│0.0, 0.0, 0.0│9615, 9604│{'pedigree_id': 36147, 'pedigree_p1':...║
║9   │196608│0.0, 0.0, 0.0│   9583, 3│{'pedigree_id': 36161, 'pedigree_p1':...║
║10  │196608│0.0, 0.0, 0.0│9609, 9583│{'pedigree_id': 36148, 'pedigree_p1':...║
║11  │196608│0.0, 0.0, 0.0│9614, 9626│{'pedigree_id': 36168, 'pedigree_p1':...║
║12  │196608│0.0, 0.0, 0.0│      8, 9│{'pedigree_id': 36188, 'pedigree_p1':...║
║13  │196608│0.0, 0.0, 0.0│     11, 4│{'pedigree_id': 36189, 'pedigree_p1':...║
║14  │196608│0.0, 0.0, 0.0│  26, 9603│{'pedigree_id': 36209, 'pedigree_p1':...║
║15  │196608│0.0, 0.0, 0.0│    25, 24│{'pedigree_id': 36214, 'pedigree_p1':...║
║16  │196608│0.0, 0.0, 0.0│  29, 9612│{'pedigree_id': 36215, 'pedigree_p1':...║
║17  │196608│0.0, 0.0, 0.0│     5, 15│{'pedigree_id': 36230, 'pedigree_p1':...║
║18  │196608│0.0, 0.0, 0.0│     8, 16│{'pedigree_id': 36218, 'pedigree_p1':...║
║19  │196608│0.0, 0.0, 0.0│    31, 23│{'pedigree_id': 36238, 'pedigree_p1':...║
║                 9588 rows skipped (tskit.set_print_options)                 ║
║9608│131072│0.0, 0.0, 0.0│9587, 9565│{'pedigree_id': 36043, 'pedigree_p1':...║
║9609│131072│0.0, 0.0, 0.0│9558, 9590│{'pedigree_id': 36058, 'pedigree_p1':...║
║9610│131072│0.0, 0.0, 0.0│9570, 9601│{'pedigree_id': 36063, 'pedigree_p1':...║
║9611│131072│0.0, 0.0, 0.0│9590, 9559│{'pedigree_id': 36048, 'pedigree_p1':...║
║9612│131072│0.0, 0.0, 0.0│9581, 9600│{'pedigree_id': 36061, 'pedigree_p1':...║
║9613│131072│0.0, 0.0, 0.0│9571, 9559│{'pedigree_id': 36055, 'pedigree_p1':...║
║9614│131072│0.0, 0.0, 0.0│9591, 9558│{'pedigree_id': 36059, 'pedigree_p1':...║
║9615│131072│0.0, 0.0, 0.0│9586, 9610│{'pedigree_id': 36074, 'pedigree_p1':...║
║9616│131072│0.0, 0.0, 0.0│9596, 9596│{'pedigree_id': 36071, 'pedigree_p1':...║
║9617│131072│0.0, 0.0, 0.0│9564, 9564│{'pedigree_id': 36080, 'pedigree_p1':...║
║9618│131072│0.0, 0.0, 0.0│9579, 9612│{'pedigree_id': 36088, 'pedigree_p1':...║
║9619│131072│0.0, 0.0, 0.0│9591, 9613│{'pedigree_id': 36098, 'pedigree_p1':...║
║9620│131072│0.0, 0.0, 0.0│9610, 9608│{'pedigree_id': 36085, 'pedigree_p1':...║
║9621│131072│0.0, 0.0, 0.0│9575, 9583│{'pedigree_id': 36111, 'pedigree_p1':...║
║9622│131072│0.0, 0.0, 0.0│9609, 9609│{'pedigree_id': 36103, 'pedigree_p1':...║
║9623│131072│0.0, 0.0, 0.0│  9603, 22│{'pedigree_id': 36126, 'pedigree_p1':...║
║9624│131072│0.0, 0.0, 0.0│  25, 9599│{'pedigree_id': 36129, 'pedigree_p1':...║
║9625│131072│0.0, 0.0, 0.0│9598, 9613│{'pedigree_id': 36135, 'pedigree_p1':...║
║9626│131072│0.0, 0.0, 0.0│    22, 22│{'pedigree_id': 36131, 'pedigree_p1':...║
║9627│131072│0.0, 0.0, 0.0│9602, 9599│{'pedigree_id': 36125, 'pedigree_p1':...║
╚════╧══════╧═════════════╧══════════╧════════════════════════════════════════╝

nodes
╔═════╤═════╤══════════╤══════════╤════╤════════════════════════════════════════╗
║id   │flags│population│individual│time│metadata                                ║
╠═════╪═════╪══════════╪══════════╪════╪════════════════════════════════════════╣
║0    │    1│         1│        38│ 999│{'slim_id': 14, 'is_null': False, 'ge...║
║1    │    1│         1│        38│ 999│{'slim_id': 15, 'is_null': False, 'ge...║
║2    │    1│         1│        39│ 999│{'slim_id': 2, 'is_null': False, 'gen...║
║3    │    1│         1│        39│ 999│{'slim_id': 3, 'is_null': False, 'gen...║
║4    │    1│         1│        40│ 999│{'slim_id': 4, 'is_null': False, 'gen...║
║5    │    1│         1│        40│ 999│{'slim_id': 5, 'is_null': False, 'gen...║
║6    │    1│         1│        41│ 999│{'slim_id': 10, 'is_null': False, 'ge...║
║7    │    1│         1│        41│ 999│{'slim_id': 11, 'is_null': False, 'ge...║
║8    │    1│         2│        42│ 999│{'slim_id': 40, 'is_null': False, 'ge...║
║9    │    1│         2│        42│ 999│{'slim_id': 41, 'is_null': False, 'ge...║
║10   │    1│         2│        43│ 999│{'slim_id': 52, 'is_null': False, 'ge...║
║11   │    1│         2│        43│ 999│{'slim_id': 53, 'is_null': False, 'ge...║
║12   │    1│         1│        44│ 999│{'slim_id': 0, 'is_null': False, 'gen...║
║13   │    1│         1│        44│ 999│{'slim_id': 1, 'is_null': False, 'gen...║
║14   │    1│         2│        45│ 999│{'slim_id': 70, 'is_null': False, 'ge...║
║15   │    1│         2│        45│ 999│{'slim_id': 71, 'is_null': False, 'ge...║
║16   │    1│         2│        46│ 999│{'slim_id': 56, 'is_null': False, 'ge...║
║17   │    1│         2│        46│ 999│{'slim_id': 57, 'is_null': False, 'ge...║
║18   │    1│         2│        47│ 999│{'slim_id': 58, 'is_null': False, 'ge...║
║19   │    1│         2│        47│ 999│{'slim_id': 59, 'is_null': False, 'ge...║
║                  19216 rows skipped (tskit.set_print_options)                 ║
║19236│    1│         1│        17│   0│{'slim_id': 72460, 'is_null': False, ...║
║19237│    1│         1│        17│   0│{'slim_id': 72461, 'is_null': False, ...║
║19238│    1│         1│        18│   0│{'slim_id': 72436, 'is_null': False, ...║
║19239│    1│         1│        18│   0│{'slim_id': 72437, 'is_null': False, ...║
║19240│    1│         2│        19│   0│{'slim_id': 72476, 'is_null': False, ...║
║19241│    1│         2│        19│   0│{'slim_id': 72477, 'is_null': False, ...║
║19242│    1│         1│        20│   0│{'slim_id': 72458, 'is_null': False, ...║
║19243│    1│         1│        20│   0│{'slim_id': 72459, 'is_null': False, ...║
║19244│    1│         1│        21│   0│{'slim_id': 72466, 'is_null': False, ...║
║19245│    1│         1│        21│   0│{'slim_id': 72467, 'is_null': False, ...║
║19246│    1│         2│        33│   0│{'slim_id': 72496, 'is_null': False, ...║
║19247│    1│         2│        33│   0│{'slim_id': 72497, 'is_null': False, ...║
║19248│    1│         2│        34│   0│{'slim_id': 72484, 'is_null': False, ...║
║19249│    1│         2│        34│   0│{'slim_id': 72485, 'is_null': False, ...║
║19250│    1│         2│        35│   0│{'slim_id': 72500, 'is_null': False, ...║
║19251│    1│         2│        35│   0│{'slim_id': 72501, 'is_null': False, ...║
║19252│    1│         2│        36│   0│{'slim_id': 72478, 'is_null': False, ...║
║19253│    1│         2│        36│   0│{'slim_id': 72479, 'is_null': False, ...║
║19254│    1│         2│        37│   0│{'slim_id': 72494, 'is_null': False, ...║
║19255│    1│         2│        37│   0│{'slim_id': 72495, 'is_null': False, ...║
╚═════╧═════╧══════════╧══════════╧════╧════════════════════════════════════════╝

**Reply by petrelharp** [reply]
Oh, good! Hopefully you can use sgkit? I'm not familiar, but I'm sure *someone* has implemented those recursive algorithms you mention.

**Reply by jeromekelleher** [reply]
Great! Would you mind attaching this trees file here please @JimmyPeniston? I'll see if I can do the tskit->sgkit linkage and post the code.

**Reply by JimmyPeniston** [reply]
Thank you @jeromekelleher! You can [download the .trees file here](https://github.com/JimmyPeniston/Pedigree_Inbreeding/blob/main/PIB_example.trees). (It wasn't letting me attached .trees file on this forum)

**Reply by jeromekelleher** [reply]
Great, thanks @JimmyPeniston! There are two issues with this:

1. All of the individuals seem to samples (we have 9628 individuals, and they are all associated with sample nodes). Is there some set of individuals you want to compute the inbreeding coefficient for?
2. This is quite a large example - do you think you could but it down to something that with a few hundred individuals (at most) please?

**Reply by JimmyPeniston** [reply]
Sorry, the number of individuals quickly got large when I started tracking all individuals that ever lived. Here is a [smaller data set with 57 individuals](https://github.com/JimmyPeniston/Pedigree_Inbreeding/blob/main/PIB_example.trees).

My end goal is to calculate inbreeding coefficients for a random sample of extant individuals in each subpopulation.

**Reply by timothymillar** [reply]
I had a quick go at this. I'm not that familiar with tskit so hopefully my use interpretation is correct.

This makes minimal assumptions about the tskit identifiers and converts them to strings:
```python
import numpy as np
import xarray as xr
import tskit as ts
import sgkit as sg

def tskit_to_sgkit_ped(trees):
    n = trees.num_individuals
    sample_id = np.zeros((n), int)
    parent_id = np.full((n, 2), -1, int)  # use -1 to indicate unknown parent
    for i, ind in enumerate(trees.individuals()):
        sample_id[i] = ind.id
        for j, p in enumerate(ind.parents):
            parent_id[i, j] = p
    ds = xr.Dataset()
    # convert ids to strings using "." to indicate unknown parent (default in sgkit)
    sample_id = sample_id.astype(str)
    parent_id = np.where(parent_id < 0, ".", parent_id.astype(str))
    ds["sample_id"] = "samples", sample_id
    ds["parent_id"] = ["samples", "parents"], parent_id
    return ds

trees = ts.load("PIB_example.trees")
ds = tskit_to_sgkit_ped(trees)
ds = sg.pedigree_inbreeding(ds)
ds.stat_pedigree_inbreeding.values
```
gives:
```
array([0.        , 0.734375  , 0.31347656, 0.2890625 , 0.33007812,
       0.36358643, 0.66503906, 0.41113281, 0.40527344, 0.78125   ,
       0.31445312, 0.734375  , 0.35137939, 0.33935547, 0.31640625,
       0.43212891, 0.        , 0.        , 0.        , 0.5       ,
       0.5       , 0.        , 0.5       , 0.25      , 0.5       ,
       0.5       , 0.5       , 0.5625    , 0.75      , 0.25      ,
       0.5625    , 0.        , 0.625     , 0.625     , 0.        ,
       0.1875    , 0.75      , 0.75      , 0.6875    , 0.375     ,
       0.3203125 , 0.40625   , 0.6875    , 0.16015625, 0.71875   ,
       0.        , 0.6875    , 0.71875   , 0.31445312, 0.09375   ,
       0.31347656, 0.50537109, 0.71875   , 0.43212891, 0.6862793 ,
       0.2890625 , 0.65722656, 0.7265625 ])
```

If you can assume that the tskit individuals are always zero-based integers (sorted in ascending order and without gaps), then you can be slightly more efficient by creating the parent indices array directly:

```python
def tskit_to_sgkit_ped(trees):
    n = trees.num_individuals
    parent = np.full((n, 2), -1, int)  # use -1 to indicate unknown parent
    for i, ind in enumerate(trees.individuals()):
        for j, p in enumerate(ind.parents):
            parent[i, j] = p
    ds = xr.Dataset()
    ds["parent"] = ["samples", "parents"], parent
    return ds
```

FWIW the sgkit code should handle 10k samples easily.

**Reply by jeromekelleher** [reply]
Thanks @timothymillar, that's exactly what I was hoping to do :heart:

**Reply by JimmyPeniston** [reply]
This is great! Thanks @timothymillar @jeromekelleher @petrelharp!!!! You have all been incredibly helpful.

**Reply by jeromekelleher** [reply]
Would you mind reporting back on how you get on using the code @JimmyPeniston? It would be useful to know how well it scales for these types of sims.

**Reply by JimmyPeniston** [reply]
Certainly! I'll report back here with more updates once I've expanded the code for what I need.
As of now, I just tried running the above code for a simulation with 10,000 stored individuals and it took less than 5 seconds.

**Reply by timothymillar** [reply]
Just a note on the `pedigree_inbreeding` implementation; It uses a (unrolled) recursive algorithm to calculate a minimal set of kinship coefficients required to get the inbreeding coefficients. The performance depends upon the complexity of the pedigree rather than just the number of samples. I haven't gone through the exact complexity notation yet but a shallower pedigree, with more parent reuse, will be much faster than a deep pedigree with random parents.
